File: Patterns/Command Pattern/Cpp/Command/CailingFan/CailingFanOnToHighCommand.cpp
#include "../ICommand.cpp"
#include "./CailingFan.cpp"

class CailingFanOnToHighCommand : public ICommand {
private:
    CailingFan* cailingFan;
    int preview;

public:
    CailingFanOnToHighCommand(CailingFan* cailingFan) : cailingFan{cailingFan}, preview{cailingFan->OFF} {}

    void execute() override {
        preview = cailingFan->getState();
        cailingFan->setHigh();
    }

    void undo() override {
        if (preview == cailingFan->MEDIUM) cailingFan->setMedium();
        else if (preview == cailingFan->LOW) cailingFan->setLow();
        else if (preview == cailingFan->OFF) cailingFan->off();
    }

    string getCommandName() override {
        return "CailingFanOnToHighCommand";
    }
};

File: Patterns/Command Pattern/CSharp/Command/CailingFan/CailingFanOnToHighCommand.cs
namespace Command.CailingFan
{
    public class CailingFanOnToHighCommand : ICommand
    {
        private CailingFan cailingFan;
        private int preview;

        public CailingFanOnToHighCommand(CailingFan cailingFan)
        {
            preview = CailingFan.OFF;
            this.cailingFan = cailingFan;
        }

        public void Execute()
        {
            preview = cailingFan.State;
            cailingFan.SetHigh();
        }

        public void Undo()
        {
            switch (preview)
            {
                case CailingFan.MEDIUM:
                    cailingFan.SetHigh();
                    break;
                case CailingFan.LOW:
                    cailingFan.SetLow();
                    break;
                case CailingFan.OFF:
                    cailingFan.Off();
                    break;
            }
        }
    }
}